python/semantic_kernel/connectors/ai/anthropic/services/anthropic_chat_completion.py

```diff
@@ -254,39 +277,154 @@ def _create_chat_message_content(
+            ]
+
+            if (
+                settings.function_choice_behavior
+                and settings.function_choice_behavior.type_ == FunctionChoiceType.REQUIRED
+            ) or type == FunctionChoiceType.REQUIRED:
+                settings.tool_choice = {"type": "any"}
+            else:
+                settings.tool_choice = {"type": type.value}
+
+    def kernel_function_metadata_to_function_call_format_anthropic(
+        self,
+        metadata: KernelFunctionMetadata,
+    ) -> dict[str, Any]:
+        """Convert the kernel function metadata to function calling format."""
+        return {
+            "name": metadata.fully_qualified_name,
+            "description": metadata.description or "",
+            "input_schema": {
+                "type": "object",
+                "properties": {p.name: p.schema_data for p in metadata.parameters},
+                "required": [p.name for p in metadata.parameters if p.is_required],
+            },
+        }
+
+    @override
+    def _update_function_choice_settings_callback(
+        self,
+    ):
+        return self.update_settings_from_function_call_configuration_anthropic
+
+    async def _send_chat_request(self, settings: AnthropicChatPromptExecutionSettings) -> list["ChatMessageContent"]:
+        """Send the chat request."""
+        try:
+            response = await self.async_client.messages.create(**settings.prepare_settings_dict())
+        except Exception as ex:
+            raise ServiceResponseException(
+                f"{type(self)} service failed to complete the request",
+                ex,
+            ) from ex
+
+        response_metadata: dict[str, Any] = {"id": response.id}
+        if hasattr(response, "usage") and response.usage is not None:
+            response_metadata["usage"] = response.usage
+
+        return [self._create_chat_message_content(response, response_metadata)]
+
+    async def _send_chat_stream_request(
+        self, settings: AnthropicChatPromptExecutionSettings
+    ) -> AsyncGenerator[list["StreamingChatMessageContent"], None]:
+        """Send the chat stream request."""
+        try:
+            async with self.async_client.messages.stream(**settings.prepare_settings_dict()) as stream:
+                metadata: dict[str, Any] = {"usage": {}, "id": None}
+                async for stream_event in stream:
+                    if isinstance(stream_event, RawMessageStartEvent):
+                        metadata["usage"]["input_tokens"] = stream_event.message.usage.input_tokens
+                        metadata["id"] = stream_event.message.id
+                    elif isinstance(stream_event, (TextEvent, RawMessageDeltaEvent)) or (
+                        isinstance(stream_event, ContentBlockStopEvent)
+                        and stream_event.content_block.type == "tool_use"
+                    ):
+                        yield [self._create_streaming_chat_message_content(stream_event, metadata)]
```

Comment:
Will the metadata get overwritten completely each time there is a `RawMessageStartEvent`? If not, then the metadata from a previous event may become the metadata of a later event.

Reply:
There are two relevant metadata pieces from `RawMessageStartEvent` (id and input tokens) and there is only 1 `RawMessageStartEvent` per response. These two pieces of metadata don't change, but the output tokens is correctly determined at the end of the message via the `RawMessageDeltaEvent`.

Example `RawMessageStartEvent` stream event:
```python
RawMessageStartEvent(message=Message(id='msg_01RBZ8YFmLbxzzKcEANCqGXv', content=[], model='claude-3-opus-20240229', role='assistant', stop_reason=None, stop_sequence=None, type='message', usage=Usage(input_tokens=1720, output_tokens=2)), type='message_start')
```

Reply:
Got it. Perhaps it's worth explicitly documenting in code that there will be only one `RawMessageStartEvent` per response (I do see now the name implies it).